runtime/src/accounts_db.rs

```diff
@@ -1984,6 +2202,211 @@ impl AccountsDB {
+            .store_append_accounts
+            .fetch_add(total_append_accounts_us, Ordering::Relaxed);
+        self.stats
+            .store_find_store
+            .fetch_add(total_storage_find_us, Ordering::Relaxed);
+        infos
+    }
+
+    pub fn mark_slot_frozen(&self, slot: Slot) {
+        if let Some(slot_cache) = self.accounts_cache.slot_cache(slot) {
+            slot_cache.mark_slot_frozen();
+            slot_cache.report_slot_store_metrics();
+        }
+        self.accounts_cache.report_size();
+    }
+
+    // Force flush the cached roots, flush any unrooted frozen slots as well if there are
+    // > MAX_CACHE_SLOTS of them.
+    pub fn force_flush_accounts_cache(&self) {
+        self.flush_accounts_cache(true);
+    }
+
+    pub fn flush_accounts_cache_if_needed(&self) {
+        self.flush_accounts_cache(false);
+    }
+
+    fn flush_accounts_cache(&self, force_flush: bool) {
+        if !force_flush && self.accounts_cache.num_slots() <= MAX_CACHE_SLOTS {
+            return;
+        }
+
+        // Flush all roots
+        let mut flush_roots_elapsed = Measure::start("flush_roots_elapsed");
+        let cached_roots = self.accounts_cache.clear_roots();
+        let mut max_root = None;
+        for root in cached_roots {
+            self.flush_slot_cache(root);
+            if let Some(max_root_slot) = max_root {
+                if root > max_root_slot {
+                    max_root = Some(root);
+                }
+            } else {
+                max_root = Some(root);
+            }
+        }
+        flush_roots_elapsed.stop();
+
+        // Remove all unrooted slots less than `max_root`, as those are no
+        // longer relevant. See comment in BankForks::set_root() to see why this is safe
+        // to do.
+        if let Some(max_root) = max_root {
+            let removed_slots = self.accounts_cache.remove_slots_le(max_root);
+            for (removed_slot, slot_cache) in removed_slots {
+                // All roots less than `max_root` should have been flushed, so
+                // we shouldn't see any here.
+                assert!(!self.accounts_index.is_root(removed_slot));
+                // Purge the state like accounts index, hashes for unrooted slots
+                let reclaims = self.purge_slot_cache_keys(removed_slot, slot_cache);
+                for (reclaimed_slot, reclaimed_account_info) in reclaims {
+                    assert_eq!(reclaimed_slot, removed_slot);
+                    assert_eq!(reclaimed_account_info.store_id, Self::cache_store_id());
+                }
+            }
+        }
+
+        // If there are > MAX_CACHE_SLOTS, then flush the excess ones to storage
+        for old_slot in self.accounts_cache.find_older_frozen_slots(MAX_CACHE_SLOTS) {
+            self.flush_slot_cache(old_slot);
+        }
+
+        inc_new_counter_info!("flush_roots_elapsed", flush_roots_elapsed.as_us() as usize);
+    }
+
+    fn flush_slot_cache(&self, slot: Slot) {
```

Comment:
nits: `slot` => `root`.

Reply:
hehe it used to be called that before I introduced the behavior to flush non-rooted slots as well.  Any frozen bank can now be flushed if there are over `MAX_CACHE_SLOTS` in the cache, checked by `flush_accounts_cache()